File: lib/capture.mjs
import path from 'node:path';
import { writeFile } from 'node:fs/promises';
import { sha256Hex, sha256File, applySubstitutions, collectFiles, runCommand } from './util.mjs';
import { appendRecord, ensureStore, ARTIFACTS_DIR } from './chain.mjs';

export async function runCapture(opts) {
  const {
    cwd = process.cwd(),
    command,
    inputs = [],
    outputs = [],
    redactions = [],
    normalizers = [],
    label = null,
    storeEnvValues = false,
  } = opts;
  if (!Array.isArray(command) || command.length === 0) {
    throw new Error('command must be a non-empty array');
  }

  await ensureStore(cwd);

  const inputFiles = [];
  for (const target of inputs) {
    for (const rel of await collectFiles(target, cwd)) {
      inputFiles.push({ path: rel, sha256: await sha256File(path.join(cwd, rel)) });
    }
  }

  const startedAt = Date.now();
  const { stdout, stderr, exitCode } = await runCommand(command, cwd);
  const durationMs = Date.now() - startedAt;

  const stdoutFinal = applySubstitutions(applySubstitutions(stdout, redactions), normalizers);
  const stderrFinal = applySubstitutions(applySubstitutions(stderr, redactions), normalizers);

  const outputFiles = [];
  for (const target of outputs) {
    for (const rel of await collectFiles(target, cwd)) {
      outputFiles.push({ path: rel, sha256: await sha256File(path.join(cwd, rel)) });
    }
  }

  const env = {};
  for (const key of Object.keys(process.env).sort()) {
    const value = String(process.env[key] ?? '');
    env[key] = storeEnvValues ? value : sha256Hex(value);
  }

  const payload = {
    label,
    command,
    cwd,
    exitCode,
    durationMs,
    stdoutHash: sha256Hex(stdoutFinal),
    stderrHash: sha256Hex(stderrFinal),
    stdoutBytes: Buffer.byteLength(stdoutFinal, 'utf8'),
    stderrBytes: Buffer.byteLength(stderrFinal, 'utf8'),
    envMode: storeEnvValues ? 'values' : 'hashes',
    env,
    redactions,
    normalizers,
    inputFiles,
    outputFiles,
    runtime: {
      platform: process.platform,
      arch: process.arch,
      node: process.versions.node,
    },
  };

  const record = await appendRecord(cwd, 'capture', payload);

  await writeFile(path.join(cwd, ARTIFACTS_DIR, `${record.id}.stdout.txt`), stdoutFinal, 'utf8');
  await writeFile(path.join(cwd, ARTIFACTS_DIR, `${record.id}.stderr.txt`), stderrFinal, 'utf8');

  return record;
}

File: cpp/capture.cpp
// capture.cpp — see capture.h.
#include "capture.h"

#include "chain.h"
#include "jsjson.h"
#include "jsre.h"
#include "sha256.h"
#include "util.h"

#include <chrono>
#include <stdexcept>

namespace pocev {

namespace {

Json fileEntry(const std::string& rel, const std::string& hash) {
    Json f = Json::object();
    f.set("path", Json::string(rel));
    f.set("sha256", Json::string(hash));
    return f;
}

long long wallMs() {
    using namespace std::chrono;
    return duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
}

const char* platformName() {
#ifdef _WIN32
    return "win32";
#elif defined(__APPLE__)
    return "darwin";
#else
    return "linux";
#endif
}

const char* archName() {
#if defined(_M_ARM64) || defined(__aarch64__)
    return "arm64";
#elif defined(_M_IX86) || defined(__i386__)
    return "ia32";
#else
    return "x64";
#endif
}

Json hashFileList(const std::vector<std::string>& targets, const std::string& cwd) {
    Json out = Json::array();
    for (const auto& target : targets) {
        for (const auto& rel : collectFiles(target, cwd)) {
            out.push(fileEntry(rel, sha256File(pathJoin(cwd, rel))));
        }
    }
    return out;
}

}  // namespace

Json runCapture(const CaptureOpts& opts) {
    if (opts.command.empty()) {
        throw std::runtime_error("command must be a non-empty array");
    }
    const std::string& cwd = opts.cwd;

    ensureStore(cwd);

    Json inputFiles = hashFileList(opts.inputs, cwd);

    long long startedAt = wallMs();
    CommandResult run = runCommand(opts.command, cwd);
    long long durationMs = wallMs() - startedAt;

    std::string stdoutFinal = applySubstitutions(applySubstitutions(run.out, opts.redactions),
                                                 opts.normalizers);
    std::string stderrFinal = applySubstitutions(applySubstitutions(run.err, opts.redactions),
                                                 opts.normalizers);

    Json outputFiles = hashFileList(opts.outputs, cwd);

    Json env = Json::object();
    for (const auto& kv : sortedEnv()) {
        env.set(kv.first,
                Json::string(opts.storeEnvValues ? kv.second : sha256Hex(kv.second)));
    }

    Json payload = Json::object();
    payload.set("label", opts.hasLabel ? Json::string(opts.label) : Json::null());
    Json command = Json::array();
    for (const auto& c : opts.command) command.push(Json::string(c));
    payload.set("command", command);
    payload.set("cwd", Json::string(cwd));
    payload.set("exitCode", run.signaled ? Json::null() : Json::number(static_cast<double>(run.code)));
    payload.set("durationMs", Json::number(static_cast<double>(durationMs)));
    payload.set("stdoutHash", Json::string(sha256Hex(stdoutFinal)));
    payload.set("stderrHash", Json::string(sha256Hex(stderrFinal)));
    payload.set("stdoutBytes", Json::number(static_cast<double>(stdoutFinal.size())));
    payload.set("stderrBytes", Json::number(static_cast<double>(stderrFinal.size())));
    payload.set("envMode", Json::string(opts.storeEnvValues ? "values" : "hashes"));
    payload.set("env", env);
    Json redactions = Json::array();
    for (const auto& r : opts.redactions) {
        Json j = Json::object();
        j.set("pattern", Json::string(r.pattern));
        j.set("flags", Json::string(r.flags));
        j.set("replacement", Json::string(r.replacement));
        redactions.push(j);
    }
    payload.set("redactions", redactions);
    Json normalizers = Json::array();
    for (const auto& r : opts.normalizers) {
        Json j = Json::object();
        j.set("pattern", Json::string(r.pattern));
        j.set("flags", Json::string(r.flags));
        j.set("replacement", Json::string(r.replacement));
        normalizers.push(j);
    }
    payload.set("normalizers", normalizers);
    payload.set("inputFiles", inputFiles);
    payload.set("outputFiles", outputFiles);
    Json runtime = Json::object();
    runtime.set("platform", Json::string(platformName()));
    runtime.set("arch", Json::string(archName()));
    runtime.set("node", Json::string("cpp"));
    payload.set("runtime", runtime);

    Json record = appendRecord(cwd, "capture", payload);

    writeFileBytes(
        pathJoin(pathJoin(cwd, ARTIFACTS_DIR), record.get("id")->str + ".stdout.txt"),
        stdoutFinal);
    writeFileBytes(
        pathJoin(pathJoin(cwd, ARTIFACTS_DIR), record.get("id")->str + ".stderr.txt"),
        stderrFinal);
    return record;
}

}  // namespace pocev
